test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -0,0 +1,648 @@
+                .builder()
+                .with(DurableLogConfig.CHECKPOINT_MIN_COMMIT_COUNT, 1)
+                .with(DurableLogConfig.CHECKPOINT_COMMIT_COUNT, 10)
+                .with(DurableLogConfig.CHECKPOINT_TOTAL_COMMIT_LENGTH, 10L * 1024 * 1024L)
+                .with(DurableLogConfig.START_RETRY_DELAY_MILLIS, 20)
+                .build();
+
+        private final ReadIndexConfig readIndexConfig = ReadIndexConfig.builder().with(ReadIndexConfig.STORAGE_READ_ALIGNMENT, 1024).build();
+        private final AttributeIndexConfig attributeIndexConfig = AttributeIndexConfig
+                .builder()
+                .with(AttributeIndexConfig.MAX_INDEX_PAGE_SIZE, 2 * 1024)
+                .with(AttributeIndexConfig.ATTRIBUTE_SEGMENT_ROLLING_SIZE, 1000)
+                .build();
+        private final WriterConfig writerConfig = WriterConfig
+                .builder()
+                .with(WriterConfig.FLUSH_THRESHOLD_BYTES, 1)
+                .with(WriterConfig.FLUSH_THRESHOLD_MILLIS, 25L)
+                .with(WriterConfig.MIN_READ_TIMEOUT_MILLIS, 10L)
+                .with(WriterConfig.MAX_READ_TIMEOUT_MILLIS, 250L)
+                .build();
+
+        DebugTool(BookKeeperLogFactory dataLogFactory, StorageFactory storageFactory) {
+            this.dataLogFactory = dataLogFactory;
+            this.storageFactory = storageFactory;
+            this.operationLogFactory = new DurableLogFactory(durableLogConfig, this.dataLogFactory, executorService);
+
+            this.cacheStorage = new DirectMemoryCache(Integer.MAX_VALUE);
+            this.cacheManager = new CacheManager(CachePolicy.INFINITE, this.cacheStorage, executorService);
+            this.readIndexFactory = new ContainerReadIndexFactory(readIndexConfig, this.cacheManager, executorService);
+            this.attributeIndexFactory = new ContainerAttributeIndexFactoryImpl(attributeIndexConfig, this.cacheManager, executorService);
+            this.writerFactory = new StorageWriterFactory(writerConfig, executorService);
+
+            ContainerConfig containerConfig = ServiceBuilderConfig.getDefaultConfig().getConfig(ContainerConfig::builder);
+            this.containerFactory = new StreamSegmentContainerFactory(containerConfig, this.operationLogFactory,
+                    this.readIndexFactory, this.attributeIndexFactory, this.writerFactory, this.storageFactory,
+                    this::createContainerExtensions, executorService);
+        }
+
+        private Map<Class<? extends SegmentContainerExtension>, SegmentContainerExtension> createContainerExtensions(
+                SegmentContainer container, ScheduledExecutorService executor) {
+            return Collections.singletonMap(ContainerTableExtension.class, new ContainerTableExtensionImpl(container, this.cacheManager, executor));
+        }
+
+        @Override
+        public void close() {
+            this.readIndexFactory.close();
```

Comment:
Verify that you are shutting down everything that needs to shut down.

Reply:
Done.